### PR TITLE
Add LoRA expert support for GPT-Neo MoE

### PR DESCRIPTION
**Pull request by ananyamjain**
## Description
This PR adds LoRA (Low-Rank Adaptation) expert support to the T-MoE framework. We implemented a LoRALayer primitive that wraps a frozen linear layer with trainable low-rank adapters, a GPTNeoLoRAMLP expert class that mirrors the GPT-Neo MLP structure, and an ExpertPool to manage multiple experts. These are wired together in LoRAMoELayer, which keeps the backbone frozen and routes tokens to LoRA experts using the metabolic router. We also added an experiment config for GPT-Neo 125M and unit/integration tests to verify that gradients only flow to the router and experts while the backbone stays frozen.

## Related Issue
Fixes https://github.com/UofT-CSC490-W2026/T-MoE/issues/11

## Type of Change
- [ ] Bug fix
- [X] New feature
- [ ] Breaking change
- [ ] Documentation update

## Testing & Verification
- [x] All tests passed (`pytest`)
- [ ] Verified on local/AWS environment
- [x] No linting errors (`ruff check .`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/experts/gpt_neo_lora.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        self.c_proj = LoRALayer(
+            config.intermediate_dim,
+            config.hidden_dim,
+            rank=config.rank,
+            alpha=config.alpha,
+            dropout=config.dropout,
+            init_scale=config.init_scale,
+        )
+        self.act = _get_activation()
+        self.dropout = nn.Dropout(config.dropout)
```
Comment (Aviral-03) [inline src/experts/gpt_neo_lora.py]: GPT-Neo's MLP does not have a dropout at the output. `GPTNeoLoRAMLP` adds one:

```python
self.dropout = nn.Dropout(config.dropout)  # not in original GPT-Neo MLP
```

At init with `dropout=0.0` this is a no-op (using `nn.Dropout(0.0)`) but if `dropout > 0` is set in config it silently deviates from the pretrained architecture. Either remove it, or document explicitly that it's an intentional regularization addition.

File: src/experts/lora.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+from dataclasses import dataclass
+from typing import Optional
+
+import torch
+from torch import nn
+
+from src.experts.base import BaseExpert
+
+
+# ── Config ──────────────────────────────────────────────────────────────────
```
Comment (Aviral-03) [inline src/experts/lora.py]: Can you remove this?

**Review by Aviral-03 — APPROVED** [review]
